### DISCUSSION TITLE
General discussion : How to regroup controller methods into controller classes ?

### DISCUSSION THREAD

**Discussion by Trismegiste** [symfony/symfony#57209] [Q&A] [answered]
Hello,
This thread is not about a specific problem but a broader design discussion : 

_In Symfony, how do you regroup a bunch of controller methods into a bunch of controller classes ?_

I mean, it's pretty obvious when you design a CRUD for an orphan entity but what about persisting a relation between 2 ou 3 entities ? What about generating a picture with GD from the content of 3 different entities in the model ? There are many use cases that don't fit into the paradigm 1-entity/1-route/1-controller/1-folder for templates.

Particularly in a big project, what's your first criterion to choose the controller class you'll append a new method into ?
* the controller already deals with the same entity ?
* the controller uses similar route attributes ?
* the controller already uses the same services from the container (already injected in the constructor) ?
* the controller already deals with the same type of Response (JSON, HTML, Picture...) ?

I think there are no good or wrong answers, feel free to explain your design decision, in fact, I'm very curious :yum: 

Thanks in advance :heart:

**Comment by MatTheCat** [comment, marked as answer]
Hello,

> There are many **use cases** that don't fit into the paradigm 1-entity/1-route/1-controller/1-folder for templates.

Then, what do you say about grouping controllers by use-case?

I’m currently a big fan of the [Vertical Slice Architecture](https://www.jimmybogard.com/vertical-slice-architecture/) where instead of grouping files by whatever pattern or technical concern, you put all files specific to a slice (aka use-case) into a single folder. Such a controller will usually have a single method, but a use-case could require more (let’s say a form-handling one and a confirmation page display e.g.).

**Reply by Trismegiste** [reply]
That's very interesting, thank you, I didn't know this is somewhat an "official" design.

It's funny because it reminds me the endless (and pointless) debate with my mother and my girlfriend on how to store things in the kitchen. I make a picture instead of a long text :arrow_down: I think the analogy is good :wink: 

This design is very interesting because I realized my E2E tests are already organized in use-cases, so why not for the rest of the code.

## Horizontal VS Vertical 
![group thing](https://github.com/symfony/symfony/assets/1260026/13a644ca-8d0b-4d95-b1c4-400865dc3a73)

**Reply by MatTheCat** [reply]
Indeed never thought about this analogy but it makes sense, it’s just missing the slices’ parts’ tight coupling: in your code changing the kettle would probably mean you have to change the tea cup and/or bag as well. This is another advantage of this architecture: it [gathers together things that change for the same reasons](https://en.wikipedia.org/wiki/Single-responsibility_principle).

Note that you would still need some shared parts if they don’t belong to a specific use-case (like cutlery!).

**Reply by Trismegiste** [reply]
> This is another advantage of this architecture

I like particularly this point (SRP). That's the big problem of many frameworks. 
'Say you add ONE field in ONE entity, you have to edit the entity class, the form class, at least one twig, translation files, unit test class, etc and therefore crawl ALL over the project in every folders.

Ok, everything is tidy and in the "right" place but sometimes, it's a pain in the :peach:
